[ENH] Use LogisticRegressionCV instead of LogisticRegression to improve Decoder performance

michellewang opened the pull request:
<!---
This is a suggested pull request template for nilearn.
It's designed to capture information we've found to be useful in reviewing pull requests.

If there is other information that would be helpful to include, please don't hesitate to add it!

Please make sure your pull request also follows the
[contribution guidelines](https://nilearn.github.io/stable/development.html#contribution-guidelines) that
will be enforced during the review process.
-->

<!-- Please indicate after the # which issue you're closing with this PR.
This is helpful for the maintainers AND will magically close the issue when this
pull request is merged!
If the PR closes multiple issues, includes "closes" before each one is listed.
https://help.github.com/articles/closing-issues-using-keywords -->
Closes #2200.

<!-- Please give a brief overview of what has changed in the PR.
If you're not sure what to write, consider it a note to the maintainers to indicate
what they should be looking for when they review the pull request. -->
Changes proposed in this pull request:

- Replace [`LogisticRegression`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) with [`LogisticRegressionCV`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegressionCV.html) in `nilearn.decoding.decoder`
  - Note that the `param_grid` should now contain `Cs` instead of `C`, meaning that any code that explicitly provides a `param_grid = {'C': [...]}` would now break. Does this mean this needs a deprecation cycle? If so how is this usually done?
    - Another option could be to change `C` to `Cs` when processing the `param_grid` (with maybe a warning)
  - I used `LogisticRegressionCV` for both `"logistic_l1"` and `"logistic_l2"`. `sklearn` also has a [`LassoCV`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html) class, but its hyperparameters are different (`alphas` instead of `Cs`), so it seems like it would be better to use `LogisticRegressionCV` for both the L1 and L2 regularization cases (to be consistent)
- Fix incorrect links to sklearn documentation in Decoder docstring (should be `LinearSVC`, not `SVC`)

bthirion commented:
Indeed, if we break the API we need to go through deprecation cycles.

michellewang commented on nilearn/decoding/decoder.py:
```diff
@@ -176,7 +178,9 @@ def _default_param_grid(estimator, X, y):
         param_grid["alphas"] = [np.geomspace(1e-3, 1e4, 8)]
-    elif isinstance(estimator, (LogisticRegression, LinearSVC, SVR)):
+    elif isinstance(estimator, LogisticRegressionCV):
+        param_grid["Cs"] = [np.geomspace(2e-3, 2e4, 8) * min_c]
```
@bthirion for your suggested change, the reason I used `2e-3`/`2e4` instead of `1e-3`/`1e4` was to be consistent with [the previous default hyperparameter values](https://github.com/nilearn/nilearn/blob/main/nilearn/decoding/decoder.py#L179), which used `[2, 20, 200] * min_c`. 

From what I understand, the `[2, 20, 200]` is multiplied by the `min_c` value which is set to 0.5 unless the estimator uses L1 penalty, in which case `min_c` is computed with [`sklearn.svm.l1_min_c`](https://scikit-learn.org/stable/modules/generated/sklearn.svm.l1_min_c.html). So the default (for the L2 regression case) was in fact `[1, 10, 100]`, just obtained in a more convoluted way. For the L1 penalty case, the `min_c` is obtained in a more data-driven way, though it's unclear to me why it's multiplied by `[2, 20, 200]` instead of something like `[1, 10, 100]`.

bthirion replied:
Maybe adding a slight comment on this in the code is useful for future reference.
Normally, the grid should be such that we should not have to care about what exact grid was used, and I don't really have an opinion on this.

michellewang replied:
Ok, I added a comment, hopefully it makes things clearer

michellewang commented:
I have added a private helper function `_replace_param_grid_key` to replace `"C"` by `"Cs"` in the `param_grid`. I also added a `DeprecationWarning` though I'm not sure if we want to deprecate the `"C"` parameter in a future version. I have also not specified a version number because I am not sure how long deprecation cycles are typically.

As a side note, I noticed that `DeprecationWarning`s are not shown in Jupyter notebooks ([it looks like they are only shown in the `__main__` module](https://peps.python.org/pep-0565/)). I assume that this is known and just wanted to check that this isn't considered an issue.

bthirion reviewed (approved):
LGTM, thx !

jeromedockes commented:
It will probably not affect performance with the current solver, because according to [the documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegressionCV.html) there is no warm-start when using `liblinear` -- but that would have to be checked. also it uses a fixed grid of values for `C`, whereas the LassoCV has a better way of choosing values for `alpha`. so in a future PR it might still be worth adding the `LassoCV` as a possible estimator

bthirion commented:
Will merge tomorrow unless I hear otherwise. Thx !

ymzayek reviewed (approved):
LGTM thanks @michellewang
